### ISSUE TITLE
Help needed to check the revamped Java FX tutorial

### ISSUE THREAD

**Issue by damithc** [se-edu/guides#75]
Please have a look at the significantly-revamped Java FX tutorial [here](https://se-education.org/guides/tutorials/javaFx.html) and report any problems or suggestions as separate issues in this issue tracker. If you found no issues, you can reply in this thread so that your effort doesn't go unnoticed.

Strongly suggested to follow the tutorial as if you are a student (i.e., actually do the steps given) so that we can tease out hidden issues.

**Comment by jedkohjk** [comment]
LGTM! Just a few nits

![Screenshot](https://github.com/user-attachments/assets/fbe2b253-9248-4d97-8532-d97c6e03b9cc)

Grammar:

> You then decide where to layout the props for every scene.

You then decide where to place the props for every scene.

> There on, it’s just a matter of pulling the curtains on your masterpiece.

Thereafter, it’s just a matter of pulling the curtains on your masterpiece.

![Screenshot](https://github.com/user-attachments/assets/c5eb990c-dd85-497e-8038-7ae40295f73b)

Punctuation:

> This tutorial takes you through the steps of building a typical Java FX application, using a chatbot application called Duke, as a running example. Given below is what the end result can look like, if you follow this tutorial until the end.

This tutorial takes you through the steps of building a typical Java FX application, using a chatbot application called Duke as a running example. Given below is what the end result can look like if you follow this tutorial until the end:

Inconsistent whether the plural of `Node` is `Nodes` or `Node`s.

Note: the rest of this page uses `Node`s, but page 2 uses `Node`. This inconsistency is seen in other elements as well (e.g. `ImageView`). Also, part 5 just uses node and nodes, not in code form.

![Screenshot](https://github.com/user-attachments/assets/78c923f3-5bda-49ee-ad69-a84b2e66ff77)

Grammar:

> Can you have more than one `Stage` an application?

Can you have more than one `Stage` in an application?

![Screenshot](https://github.com/user-attachments/assets/16800686-5dc9-47c9-84d3-69470275e6e0)

> As customary, ...

As is customary, ...

![Screenshot](https://github.com/user-attachments/assets/119d76e0-497d-4dba-b601-12101dd936cb)

> Then, create an example DialogBox with a simple message.

Then, create an example `DialogBox` with a simple message.

![Screenshot](https://github.com/user-attachments/assets/971dfbb1-6a0e-41b7-a98d-304dc5f23c01)

Grammar:

> Follow the same for similar cases of using `getResourceAsStream` method in later parts of this tutorial.

Do the same for similar cases of using `getResourceAsStream` method in later parts of this tutorial.

![Screenshot](https://github.com/user-attachments/assets/ea87e5a3-4267-47ac-aa63-1117b5e39ab8)

> First, let's delete the following two lines that shows a dialog box by default.

First, let's delete the following two lines that show a dialog box by default:

![Screenshot](https://github.com/user-attachments/assets/883268bc-cedf-4f38-a3cc-bf0d54e85807)

> Next, let's create an instance of this Duke object in the `Main` (to be used for generating responses to user input).

Next, let's create an instance of this Duke object in the `Main` method (to be used for generating responses to user input).

![Screenshot](https://github.com/user-attachments/assets/2453ba40-bde3-4548-ba16-0581e79327c0)

> Add padding between each DialogBox
> Add padding between each ImageView and its Label
> Clip the ImageView into a circle

Add padding between each `DialogBox`
Add padding between each `ImageView` and its `Label`
Clip the `ImageView` into a circle

![Screenshot](https://github.com/user-attachments/assets/8e785e50-ef32-4737-b536-624e7a019c18)

> For example consider the following FXML snippet that defines a TextField similar to the one that we programmatically defined previous in part 2.:

For example consider the following FXML snippet that defines a `TextField` similar to the one that we programmatically defined previous in part 2:

> Once we switch over to using FXML, we can use a tool such as Scene Builder (a tool developed by Oracle and currently maintained by Gluon) to inspect and tweak the GUI design in a WYSIWYG) fashion.

Once we switch over to using FXML, we can use a tool such as Scene Builder (a tool developed by Oracle and currently maintained by Gluon) to inspect and tweak the GUI design in a WYSIWYG fashion.

![Screenshot](https://github.com/user-attachments/assets/c2c29162-f65a-4181-bc31-60c7be2f4c3a)

Grammar:

> Without setting `minHeight` to **-Infinity**, text overrun may occur when the text to be displayed exceeds the size of the label, causing the text to not be fully displayed and ending with `...` instead.

 Without setting `minHeight` to **-Infinity**, text overrun may occur when the text to be displayed exceeds the size of the label, causing the text to not be fully displayed and end with `...` instead.

![Screenshot](https://github.com/user-attachments/assets/770f1fb3-fdb6-40a1-ae2a-3c4d5a7b79ca)

> When we create a new instance of `DialogBox` (i.e., in the `DialogBox` constructor), we set both the controller and root Node to `DialogBox`.

When we create a new instance of `DialogBox` (i.e., in the `DialogBox` constructor), we set both the controller and root `Node` to `DialogBox`.

![Screenshot](https://github.com/user-attachments/assets/471f9016-6be6-4144-99ff-18ffcdb14b17)

Only part 5 shows last updated

![Screenshot](https://github.com/user-attachments/assets/bc8839ec-5383-45ea-9e41-c6cad433f453)

Grammar:

> As you may have noticed, some of the elements does not automatically follow the dimension of the app when resizing:

As you may have noticed, some of the elements do not automatically follow the dimension of the app when resizing:

> `Send Button` does not follow to bottom right of the app
> `ScrollPane` does not automatically resize both horizontally and vertically

`Send Button` does not follow the bottom right of the app
`ScrollPane` does not automatically resize either horizontally or vertically

![Screenshot](https://github.com/user-attachments/assets/f24f2924-8480-4753-aea2-7e3248133599)

> To rectify this, we can use a mixture of the following two techniques:

To rectify this, we can use a combination of the following two techniques:

Grammar:

> [Anchor `TextField`](https://se-education.org/guides/tutorials/javaFxPart5.html#anchor-textfield) to the bottom while also automatically resize horizontally

[Anchor `TextField`](https://se-education.org/guides/tutorials/javaFxPart5.html#anchor-textfield) to the bottom while also automatically resizing horizontally

![Screenshot](https://github.com/user-attachments/assets/7f09924f-f0b4-4ab0-9ae8-f1556d6cc558)

Grammar:

> At the same time we **enable Fit To Width** so that the content within the `ScrollPane` resize as well.

At the same time we **enable Fit To Width** so that the contents within the `ScrollPane` resize as well.

![Screenshot](https://github.com/user-attachments/assets/4c179760-4ee2-4d32-8689-17ec0319ac06)

> Labels, TextBoxes, Buttons, more or less all the nodes used so far are subclasses of [Region](https://docs.oracle.com/javafx/2/api/javafx/scene/doc-files/cssref.html#region), and hence share the following CSS properties for borders:

`Labels`, `TextBoxes`, `Buttons`, more or less all the nodes used so far are subclasses of [`Region`](https://docs.oracle.com/javafx/2/api/javafx/scene/doc-files/cssref.html#region), and hence share the following CSS properties for borders:

![Screenshot](https://github.com/user-attachments/assets/ff229ceb-2666-49ce-a3ce-a962f002d07f)

> Using a [ladder](https://se-education.org/guides/tutorials/javaFxPart5.html#ladder) to change color based on background
> Using [derive](https://se-education.org/guides/tutorials/javaFxPart5.html#derive) to get a brighter or darker version of selected color

Using a [ladder](https://se-education.org/guides/tutorials/javaFxPart5.html#ladder) to change color based on the background
Using [derive](https://se-education.org/guides/tutorials/javaFxPart5.html#derive) to get a brighter or darker version of 
a selected color

![Screenshot](https://github.com/user-attachments/assets/508e247d-aef3-4a19-a47d-ac59463ee6c3)

> This works for TextFields/Labels/Images as well.

This works for `TextFields`/`Labels`/`Images` as well.

> `:focused`, when the node is selected (e.g., selecting the TextField)

`:focused`, when the node is selected (e.g., selecting the `TextField`)

> Since an ImageView is a [node](https://docs.oracle.com/javafx/2/api/javafx/scene/doc-files/cssref.html#node), we can play around with the size, opacity, rotation, and more. This also means that the following properties can also be used for other kind of nodes.

Since an `ImageView` is a [node](https://docs.oracle.com/javafx/2/api/javafx/scene/doc-files/cssref.html#node), we can play around with the size, opacity, rotation, and more. This also means that the following properties can also be used for other kind of nodes.

![Screenshot](https://github.com/user-attachments/assets/4a5e26ad-121a-4075-aeec-fdaa48001698)

> [Create your own style class and the style class manually](https://se-education.org/guides/tutorials/javaFxPart5.html#flipped-text-bubbles)

[Create your own style class manually](https://se-education.org/guides/tutorials/javaFxPart5.html#flipped-text-bubbles)

![Screenshot](https://github.com/user-attachments/assets/b2b435aa-b4f0-4753-aa9d-3dd9868b9df2)

Grammar:

> Create style classes for each type of commands in `dialog-box.css`.

Create style classes for each type of command in `dialog-box.css`.

![Screenshot](https://github.com/user-attachments/assets/426ec464-3f47-48d8-a827-fd7f6b898381)

> We can set an image using CSS.

We can set a background image using CSS.

![Screenshot](https://github.com/user-attachments/assets/42bceacf-e364-4449-8b2b-fbe5ec89459a)

Grammar:

> Pictures repeating on both axis.

Pictures repeating on both axes.

![Screenshot](https://github.com/user-attachments/assets/49034943-5323-47a7-9d3c-2021a9128ba3)

> Set GIF as profile pictures

Set profile pictures as GIFs

Other general comments:

- Inconsistent on whether to put `:` or `.` after a sentence with "as shown below" or "as follows".
- Awkward line spacing in Exercises

Otherwise, LGTM!

**Comment by damithc** [comment]
@jedkohjk thanks very much for the detailed feedback. If you have time, please review https://github.com/se-edu/guides/pull/76 as well.

**Comment by jyue487** [comment]
Question:
Part 1
1. Is the correct JDK version mentioned in this tutorial?
<img width="1511" alt="Screenshot 2024-08-13 at 6 40 31 PM" src="https://github.com/user-attachments/assets/624eec16-64c3-48b9-ab43-96c5e9d29a58">
The given guide link just mentioned some examples, is it necessary that we ask them to select a particular version of JDK version?

Part 1
2. I am not sure whether we should just leave it as jdk-11 instead of other jdk versions
<img width="794" alt="Screenshot 2024-08-13 at 6 56 52 PM" src="https://github.com/user-attachments/assets/ea3a6a28-d2aa-4774-af75-1565d605a76e">
`./gradlew clean build` does not work until I commented out this line of code. Commenting this line out had made `./gradlew clean build` successful.

Part 1
3. Perhaps providing a more detailed guide on running `Launcher` would be clearer to students. For example, run `Launcher` by executing `./gradlew run`. 
<img width="539" alt="Screenshot 2024-08-13 at 7 14 01 PM" src="https://github.com/user-attachments/assets/7b2debc7-c966-43d8-bc78-e5d93a6c59cd">

Part 2
4. Perhaps a semicolon is missing here I think.
<img width="1020" alt="Screenshot 2024-08-13 at 7 28 04 PM" src="https://github.com/user-attachments/assets/7bb92eee-2cd9-4d99-bbe7-aa632181a9f3">

Part 3
5. `user` is changed to `userImage` in one of the PR's if I remembered correctly.
<img width="1248" alt="Screenshot 2024-08-13 at 7 36 48 PM" src="https://github.com/user-attachments/assets/05245856-0354-4f6c-9e6d-23ade20039a2">

This are all the issues I have found, no major issues currently.

**Comment by damithc** [comment]
> 2. I am not sure whether we should just leave it as jdk-11 instead of other jdk versions <img alt="Screenshot 2024-08-13 at 6 56 52 PM" width="794" src="https://private-user-images.githubusercontent.com/115721366/357383033-ea3a6a28-d2aa-4774-af75-1565d605a76e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjM1NjMyMTksIm5iZiI6MTcyMzU2MjkxOSwicGF0aCI6Ii8xMTU3MjEzNjYvMzU3MzgzMDMzLWVhM2E2YTI4LWQyYWEtNDc3NC1hZjc1LTE1NjVkNjA1YTc2ZS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwODEzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDgxM1QxNTI4MzlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hMWM4YjU1NjBiZWI1MGRmNTBhZDRhNzk2YTRkNjIzODY1NWVmNDA1M2RlYjRiYjYwNjVhMDg5ZmZmOWFjOTUyJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.rctyyhdbU23OGB-QnSoWMLrsZe_AWIomvAQFfhGDQiQ"> `./gradlew clean build` does not work until I commented out this line of code. Commenting this line out had made `./gradlew clean build` successful.

@jyue487 good catch. This appears to be a stray file. Removed.

**Comment by damithc** [comment]
@jyue487 Thanks for the review. I've fixed problems 1,3,4,5 as well.

**Comment by JiaXinEu** [comment]
Ran through the tutorial up to Part 4 and the code works well. Here are some minor issues:

Part 2.
```
`Scrollpane` contains one node, whereas a `Pane` can contain multiple nodes.
```
`Scrollpane` to `ScrollPane`

Part 3.
```
In part 3, we achieved the desired layout.
```
In part 2, we achieved the desired layout.


```
        dialogContainer.getChildren().addAll(new DialogBox(userInput.getText(), user));
```
dialogContainer.getChildren().addAll(new DialogBox(userInput.getText(), userImage));


```
On top of showing what the user sent, we need to take the response generated my the program and pass it to the UI components.
```
On top of showing what the user sent, we need to take the response generated by the program and pass it to the UI components.

Part 4.
```
text overrun may occur when the text to be displayed exceeds the size of the label, causing the text to not be fully displayed and ending with `...` instead:
```
text overrun may occur when the text to be displayed exceeds the size of the label, causing the text to not be fully displayed and ends with `...` instead:

The issues mentioned above are fixed in PR #78. The other two below may require further review.

Part 2.
```
Then, create an example DialogBox with a simple message.
```
Would it be clearer to say `Then, create a sample DialogBox with a simple message` instead.

Part 4.
```
Let's return to Duke and convert it to use FXML instead, with give Scene Builder a try as well.
```
The second part of the sentence seems unclear. Can it be `we will give Scene Builder a try as well`?

**Comment by damithc** [comment]
@JiaXinEu 

> Would it be clearer to say `Then, create a sample DialogBox with a simple message` instead.

Sure. We can change it in that way.



> The second part of the sentence seems unclear. Can it be `we will give Scene Builder a try as well`?

`Let's return to Duke and convert it to use FXML instead, and give Scene Builder a try as well.`?
Your suggestion is fine too.

**Comment by JiaXinEu** [comment]
Thank you for the review @damithc. 


> Let's return to Duke and convert it to use FXML instead, and give Scene Builder a try as well.

This sounds great to me. Committed the changes into PR #78

**Comment by macareonie** [comment]
Ran through the full tutorial other than some sections in Part 5. Here are just a few minor issues/nits:


## Part 3, Iteration 3 section

Last code block on the page (`Main.java`)
![image](https://github.com/user-attachments/assets/8579abdf-ceea-4706-950c-f85980127e1b)

> `String dukeText = getResponse(userInput.getText());`

`String dukeText = duke.getResponse(userInput.getText());`

## Part 5

![image](https://github.com/user-attachments/assets/25a7f698-c4eb-4784-b4ef-ca3fce859c94)

Personally, the word "stuff" feels somewhat informal or casual in a tutorial. Could possibly consider using "aspects", "elements" or "components" instead.

Grammar:
> Therefore, feel free to only look at the sections that interests you.

Therefore, feel free to only look at the sections that interest you.

## Overall:
![image](https://github.com/user-attachments/assets/289a3fbd-929d-40d6-ab16-327b48a9c6eb) 
![image](https://github.com/user-attachments/assets/7871697e-dfaf-4b19-82cd-3320f9bed71b)

The spacing between the points look off. Additionally, inconsistent use of bullet points or numbering across the tutorial parts.


![image](https://github.com/user-attachments/assets/220cd555-09ae-46a2-a04c-eb4076413206)
![image](https://github.com/user-attachments/assets/bb4229be-f0af-4b28-a6c4-57401fbc5266)
![image](https://github.com/user-attachments/assets/30933686-3ff9-4c1b-b76b-93fe69b07d7e)

Formatting of terms like "send button" and " enter key" are inconsistent.

Grammar:
> does not follow to bottom right of the app

does not follow to the bottom right of the app

<hr>

Otherwise, LGTM!

**Comment by damithc** [comment]
Thanks for the review @macareonie 
<kbd>Enter</kbd> is a key in the keyboard while `Send` is a button in the GUI (hence, different formatting). Fixed the rest. Pls check and see if I missed anything.

**Comment by macareonie** [comment]
Thanks for the clarification @damithc 

Just looked through the changes and it LGTM!

**Comment by drustanyjt** [comment]
OS: Ubuntu 22.04.4 LTS
JDK: OpenJDK 17
IDE: IntelliJ IDEA Ultimate 2024.1.4

Part 1:
1. I noticed that when first creating `Main.java` after editing the `build.gradle` file, the IDE will highlight many unresolved symbols
![image](https://github.com/user-attachments/assets/441e1fd6-d474-425f-afd9-2da04745d426)
these do not go away on using `./gradlew clean build`, but do when clicking on the refresh icon at the top right of the screenshot (I assume this loads in the new dependencies). I don't really think it's a problem, but would be helpful to keep in mind in case students ask about this error.

2. Initially I could not use `./gradlew` and had to run `chmod +x gradlew`. Is this intended? Could be an OS specific issue?

3. For reference, on my OS the hello world app looks like this 
![image](https://github.com/user-attachments/assets/b5b2639f-1345-4e6d-b2bf-881a50e2dbf8)

Part 4:
4. Nothing noteworthy but I wanted to confirm that scenebuilder works on my OS
![image](https://github.com/user-attachments/assets/9b92e36a-9f6e-4620-8d86-3a2a62c10f3a)

Part 5:
5. I tried the CSS section and while I verified that scenebuilder did update the `.fxml` file with the stylesheet path, the app it self did not appear different
![image](https://github.com/user-attachments/assets/d70a946a-f438-4a61-bb91-a53003b8e74b)

**Comment by damithc** [comment]
> 2\. Initially I could not use `./gradlew` and had to run `chmod +x gradlew`. Is this intended? Could be an OS specific issue?

Good catch, @drustanyjt 
I changed permissions of that file.
Can you clone again and try again?

**Comment by drustanyjt** [comment]
@damithc seems to be fixed, I can use `./gradlew` without issues now.
![image](https://github.com/user-attachments/assets/9b5014ae-90f7-4e19-aec6-860d188bc08d)

**Comment by billyhoce** [comment]
I followed the tutorial through Parts 1 to 4, and here are some of my findings. The findings which I feel are more important have been marked with exclamation marks :exclamation:. Overall, the tutorial has been improved greatly from how I remembered it being like, it is much easier to follow now, not to mention some of the cool animations (like resizing of the GUI window in part 5). So good job to everyone involved! For part 5, I might look through it another day.

## Part 1 -> Exercises
<img src="https://github.com/user-attachments/assets/29dc9a8c-4f17-40bd-8ab6-9a856442382b" width="500">

Should we consider introducing the JavaFX docs here, or providing a link to the [label class](https://openjfx.io/javadoc/17/javafx.controls/javafx/scene/control/Label.html)? The docs are only introduced in the next part (Part 2), and students trying out these exercises might be a little lost as to how to interact with the label class.

## Part 2 -> Designing the Layout :exclamation::exclamation::exclamation:
There is a mismatch between the layout diagram and the mockup UI. In the diagram, The "ImageView" and "Label" should be pointing into a "Hbox" which points into the "Vbox". Not sure if this was omitted for conciseness. 
<img src="https://github.com/user-attachments/assets/eb3c7503-e582-43b6-94a9-9bdc3771c2ef" width="350"> <img src="https://github.com/user-attachments/assets/0456a462-eefb-4dcd-9520-6d3b1c446986" width="304">

## Part 3 -> Iteration 1 – Echoing the user

> At first glance everything appears to work perfectly... You'll notice that when the chat entries fill up beyond **scroll pane's** display area, ... the ScrollPane down.

change "scroll pane's" to `ScrollPane`'s

> Verify that the scroll pane now scrolls as intended.

change "scroll pane" to `ScrollPane`

## Part 3 -> Iteration 2 – Adding dialog boxes for Duke's response

> Now, we can update the `handleUserInput()` as follows, ...

Changing to either of these should be more grammatically correct
"Now, we can update `handleUserInput()` as follows" 
"Now, we can update the `handleUserInput()` function as follows"

## Part 4 (Introduction paragraph) :exclamation::exclamation::exclamation:
> Problem 1: The `MainWindow` class attempts to do it all.

I think it should be `Main` instead of `MainWindow` to be consistent with previous parts. This can be quite confusing for students following the guide as this is the first ever mention of  `MainWindow`

## Part 4 -> Using Scene Builder to tweak the GUI 
For a better, I think it might be better if the image of the expanded controller panel should be moved to immediately after
> Let’s now open DialogBox.fxml in Scene Builder (open it the same way you opened the previous file), and expand the Controller panel on the left-side accordion.

![image](https://github.com/user-attachments/assets/c3fce7dc-5983-4265-a601-0a5c5597954d)

**Comment by damithc** [comment]
Thanks for the inputs @billyhoce 
I've fixed some issues as suggested, due to urgency. We can revert if others disagree.


> ## Part 1 -> Exercises
> <img alt="" width="500" src="https://private-user-images.githubusercontent.com/84035425/360452893-29dc9a8c-4f17-40bd-8ab6-9a856442382b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQ0NzY2OTMsIm5iZiI6MTcyNDQ3NjM5MywicGF0aCI6Ii84NDAzNTQyNS8zNjA0NTI4OTMtMjlkYzlhOGMtNGYxNy00MGJkLThhYjYtOWE4NTY0NDIzODJiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODI0VDA1MTMxM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWZiMmNlMzY1NmNiYWZlZTUwYzJkOGE5N2NiOTc2MGVlNTMwZTM0ZTdiOGJlYjVlYmNiZDZjNmY0MzgyZjI4ZDUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.YzM2VUt2k-VFIj0p85034Hi3fEHUbmHHui8HRj341X0">
> Should we consider introducing the JavaFX docs here, or providing a link to the [label class](https://openjfx.io/javadoc/17/javafx.controls/javafx/scene/control/Label.html)? The docs are only introduced in the next part (Part 2), and students trying out these exercises might be a little lost as to how to interact with the label class.

Waiting for inputs from others.



> ## Part 2 -> Designing the Layout ❗❗❗
> There is a mismatch between the layout diagram and the mockup UI. In the diagram, The "ImageView" and "Label" should be pointing into a "Hbox" which points into the "Vbox". Not sure if this was omitted for conciseness. <img alt="" width="350" src="https://private-user-images.githubusercontent.com/84035425/360554122-eb3c7503-e582-43b6-94a9-9bdc3771c2ef.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQ0NzY2OTMsIm5iZiI6MTcyNDQ3NjM5MywicGF0aCI6Ii84NDAzNTQyNS8zNjA1NTQxMjItZWIzYzc1MDMtZTU4Mi00M2I2LTk0YTktOWJkYzM3NzFjMmVmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODI0VDA1MTMxM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTAxZjIyY2EyNzI1YmMxNjliMzg2MTliYmFjMTdlNjdhMzE1YjdlYzM0ZTQ1MjE0M2QyMDU0NjRmZmE5YzZiMWYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.gpjaeESnwGcYh5hpFASa79VZD7RZ5gBt8Jzkw8qLlMY"> <img alt="" width="304" src="https://private-user-images.githubusercontent.com/84035425/360554419-0456a462-eefb-4dcd-9520-6d3b1c446986.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQ0NzY2OTMsIm5iZiI6MTcyNDQ3NjM5MywicGF0aCI6Ii84NDAzNTQyNS8zNjA1NTQ0MTktMDQ1NmE0NjItZWVmYi00ZGNkLTk1MjAtNmQzYjFjNDQ2OTg2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODI0VDA1MTMxM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTU3NDRmMmVmNTEzNTk0ZjY4N2Q0ZTE3ZTk4NjY2MTAxZTk4M2VhZDgyZmJkZmJhNzIyNmY2OTdmODFmM2Y1ZWMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.ROWE5K9MJFa9Oh6dFQtlPu1WifjtZl3HpIG6hU-Oaeo">

Added `HBox`. Please check.

> ## Part 3 -> Iteration 1 – Echoing the user
> > At first glance everything appears to work perfectly... You'll notice that when the chat entries fill up beyond **scroll pane's** display area, ... the ScrollPane down.
> 
> change "scroll pane's" to `ScrollPane`'s
> 
> > Verify that the scroll pane now scrolls as intended.
> 
> change "scroll pane" to `ScrollPane`

I think 'scroll pane' is used when referring to user POV while `ScrollPane` refers to developer/Java POV. Nevertheless, I've changed as suggested, for consistency.


> ## Part 3 -> Iteration 2 – Adding dialog boxes for Duke's response
> > Now, we can update the `handleUserInput()` as follows, ...
> 
> Changing to either of these should be more grammatically correct "Now, we can update `handleUserInput()` as follows" "Now, we can update the `handleUserInput()` function as follows"

`the` was there already. Added `method` to read as `Now, we can update the handleUserInput() method as follows`

> ## Part 4 (Introduction paragraph) ❗❗❗
> > Problem 1: The `MainWindow` class attempts to do it all.
> 
> I think it should be `Main` instead of `MainWindow` to be consistent with previous parts. This can be quite confusing for students following the guide as this is the first ever mention of `MainWindow`

Fixed.

> ## Part 4 -> Using Scene Builder to tweak the GUI
> For a better flow, I think it might be better if the image of the expanded controller panel should be moved to immediately after
> 
> > Let’s now open DialogBox.fxml in Scene Builder (open it the same way you opened the previous file), and expand the Controller panel on the left-side accordion.

Done

**Comment by damithc** [comment]
A question: In https://se-education.org/guides/tutorials/javaFxPart4.html we say that `minHeight="-Infinity"` prevents the dialog box from being clipped at a certain height. But this property is about the minimum height, so it is a bit puzzling (why minHeight? why _negative_ infinity?). Intuitively, it feels like the `maxHeight="Infinity"` should be the way to indicate "I do not want the dialog box to be limited to any height limit". 
Any thoughts?

**Comment by billyhoce** [comment]
> A question: In https://se-education.org/guides/tutorials/javaFxPart4.html we say that `minHeight="-Infinity"` prevents the dialog box from being clipped at a certain height. But this property is about the minimum height, so it is a bit puzzling (why minHeight? why _negative_ infinity?). Intuitively, it feels like the `maxHeight="Infinity"` should be the way to indicate "I do not want the dialog box to be limited to any height limit". Any thoughts?

The last comment on [this forum](https://coderanch.com/t/663799/java/JavaFXML-columns-layout-column-fixed) might give some insight. I'm not sure if there is a workaround though.

TLDR: `-Infinity` tells the label to set `minHeight` as the preferred height (Which is likely the length of the text in the label). Without specifying this, the minHeight property is set by default to  "USE_COMPUTED_SIZE" which might not necessarily give us the text in label.

**Comment by damithc** [comment]
> The last comment on [this forum](https://coderanch.com/t/663799/java/JavaFXML-columns-layout-column-fixed) might give some insight. I'm not sure if there is a workaround though.
> 
> TLDR: `-Infinity` is used as a constant to tell the label to set `minHeight` as the preferred height (Which is likely the length of the text in the label). Without specifying this, the minHeight property is set by default to "USE_COMPUTED_SIZE" which might not necessarily give us the text in label.

I see. Should be fine, then. Thanks for looking into  this, @billyhoce

**Comment by gongg21** [comment]
May be a little late on a review, but just a quick input! Generally, the tutorial exhibits good flow by first establishing a concrete idea of JavaFX applications using an analogy and end-product of Duke (nice animation) as a motivating example. Then proceeding to setting up and showing a quick example on _hello world_ was neat.

Following with a more elaborate description on the static implementation of a UI was smooth and then diving into dynamic features, i.e. user interaction, was well thought of.

Jumping into FXML to show how the code can become more concise was a good, natural addition after. However, the only issue, I feel from a student's perspective, is the way problem 1 is described:

> Problem 1: The Main class attempts to do it all. Code for visual tweaks, listeners and even utility methods are all in one file. This makes it difficult to find and make changes to existing code.

The emphasis on "difficult to find and make changes" is not quite substantiated, i.e. going through the tutorial, one could possibly think "I don't see the link between `Main` class doing everything and difficult to find and change code". Perhaps we could just add the entire code of `Main` class as a dropdown for readers to view and understand the impact more clearly?

Nonetheless, ending off with more advanced styling methods for fine-tuning GUI was appropriate. There was also a good balance between explanations of key concepts and telling the readers to refer to actual documentations via links provided. Overall, a very much improved version!
